File: Python/CircularQueue.py
class DataStructures.Queue:
    def __init__(self,ms):
        self.myqueue=[]
        self.maxsize=ms
        self.front=0
        self.rear=0
    def EnQueue(self):
        
        if len(self.myqueue)==self.maxsize:
            print("DataStructures.Queue is full . no new element can be added")
        else:
            data=int(input("Enter data to enqueue  "))
            self.myqueue.append(data)
            self.rear=(self.rear+1)%self.maxsize
            print("Data added to queue")
    def DeQueue(self):
        if len(self.myqueue)==0:
            print("DeQueue cannot be performed  ")
        else:
            
            print("Removed",self.myqueue.pop(self.front))
            self.front=(self.front+1)%self.maxsize
            
            if self.front>=self.rear:
                self.front=0
        
    def isEmpty(self):
        if len(self.myqueue)==0:
            print("DataStructures.Queue is empty")
        else:
            print("DataStructures.Queue is not empty")
    def display(self):
        print("Elements in DataStructures.Queue are")
        print(self.myqueue)
    def peek(self):
        print("The first element in the DataStructures.Queue is ",self.myqueue[self.front])
        
    def length(self):
        print("Length of DataStructures.CircularQueue is ",len(self.myqueue))


ms=int(input("Enter maximum size for cricular queue \n"))
Q=DataStructures.Queue(ms)
while True:
    print("Enter a option you want to perform ")
    op=int(input("\n1.enque \n2.dequeue \n3.display \n4.peek \n5.isEmpty \n6.length \n7.Exit \n"))
    if(op==1):
        
        Q.EnQueue()
    elif(op==2):
        Q.DeQueue()
    elif(op==3):
        Q.display()
    elif(op==4):
        Q.peek()
    elif(op==5):
        Q.isEmpty()
    elif(op==6):
        Q.length()
    elif(op==7):
        exit()
    else:
        print("enter correct option")


"""
Implementation of Circular DataStructures.Queue

Sample output:
Enter maximum size for cricular queue 
3
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
1
Enter data to enqueue  1
Data added to queue
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
1
Enter data to enqueue  2
Data added to queue
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
1
Enter data to enqueue  3
Data added to queue
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
3
Elements in DataStructures.Queue are
[1, 2, 3]
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
4
The first element in the DataStructures.Queue is  1
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
5
DataStructures.Queue is not empty
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
6
Length of DataStructures.CircularQueue is  3
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
2
Removed 1
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
2
Removed 2
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
2
Removed 3
Enter a option you want to perform 

1.enque 
2.dequeue 
3.display 
4.peek 
5.isEmpty 
6.length 
7.Exit 
7



"""

File: Java/DataStructures/CircularQueue.java
package Java.DataStructures;

import java.util.*;


public class CircularQueue {
    public static void main ( String[] args)
    {
         Scanner sc=new Scanner(System.in);
         System.out.println("Enter the size of circular queue: ");
         int n=sc.nextInt();
         Queue q=new Queue(n);
         while(true)
         {
        System.out.println("Press 1:To insert\n Press 2:To delete \n Press 3:To display \n Press 4:To Exit");
       int option=sc.nextInt();
       
      switch(option)
        {
            case 1: {System.out.println("Enter the element");
                          final int num=sc.nextInt();
                          q.insert(num);  }
                          break;
            case 2: {q.delete();}     
                          break;
            case 3:{q.display();}
                          break;
            case 4: System.exit(0);        //terminating the program                   

        }
        }

       }

    }

class Queue
{ 

   int front= -1;
   int rear= -1;
   int size;
   int [] cqueue;
   public Queue(final int size)
   {this.size=size;
     cqueue=new int[size];
   }

   // Inesrtion of element from rear in queue
  public void insert(final int data)
  {
   if((rear+1)%size==front)
       {
           System.out.println("Queue is full");
           return;

       }
    else if(rear==-1 && front==-1)
    {  rear=0;
       front=0; }
     else if(rear==(size-1) && front!=0)
     {rear=0;}

    else
    {rear=(rear+1)%size;

    }
    cqueue[rear] = data;
  }
//    deletion of element from front in queue
  public void delete()
  {
           int val;

          if(front==-1 && rear==-1)
          {
              System.out.println("Queue is empty");
              return ;
          }
          else if(front==rear)
          {  val=cqueue[front];
              front=-1;
              rear=-1;
          }
          else if(front==size-1)
          {   val=cqueue[front];
              front=0; }
          else
          {
              val=cqueue[front];
              front++;
          }
         System.out.println("deleted element is :"+val);


  }
// printing the queue
  public void display()
  {   if(front==-1)
       {
           System.out.println("Queue is Empty");
       }
       else if(front <=rear)
       {for(int i=front;i<=rear;i++)
           {
               System.out.print(cqueue[i]+" ");
           }

       }
       else if(front>rear)
       {for(int i=front;i<=size-1;i++)
           {
               System.out.print(cqueue[i]+" ");
           }
         for(int i=0;i<=rear;i++)
         {
             System.out.print(cqueue[i]+" ");
         }

        }



 }

}

/* Operations on Circular Queue
Sample I/O
Input
Size of circular queue:
Select the command: press 1 to insert element
                    press 2 to delete element
                    press 3 to diplay the queue
                    press 4 to exit from code

Output
display of queue

Example 1:Suppose the size of the queue is 3
           insert(3);
           insert(4);
           insert(7);
  if you again hit insert command like
           insert(5);

  Output is =>Queue is full

  if you hit display command the only starting 3 element in queue is shown
  Output is=>3 4 7

  if you hit delete command then the last element to inserted , will get delete
  Output is=>3 4

  to exit from the code hit 4 to exit.



Time Complexity:O(1)
Space Complexity:O(1)

*/
